eth/beacon/db/chain.py

```diff
@@ -0,0 +1,450 @@
+from abc import ABC, abstractmethod
+import functools
+
+from typing import (
+    Dict,
+    Iterable,
+    Tuple,
+)
+from cytoolz import (
+    first,
+    sliding_window,
+)
+
+import rlp
+from eth_utils import (
+    encode_hex,
+    to_tuple,
+    ValidationError,
+)
+from eth_typing import (
+    Hash32,
+)
+
+from eth.db.backends.base import (
+    BaseAtomicDB,
+    BaseDB,
+)
+from eth.constants import (
+    GENESIS_PARENT_HASH,
+)
+from eth.exceptions import (
+    BlockNotFound,
+    CanonicalHeadNotFound,
+    ParentNotFound,
+)
+from eth.validation import (
+    validate_block_number,
+    validate_word,
+)
+
+from eth.beacon.db.schema import SchemaV1
+
+from eth.beacon.types.block import BaseBeaconBlock  # noqa: F401
+
+
+class BaseBeaconChainDB(ABC):
+    db = None  # type: BaseAtomicDB
+
+    # FIXME: Define more abstractmethod
+
+    @abstractmethod
+    def __init__(self, db: BaseAtomicDB) -> None:
+        raise NotImplementedError("BeaconChainDB classes must implement this method")
+
+    #
+    # Block API
+    #
+    @abstractmethod
+    def persist_block(self,
+                      block: BaseBeaconBlock) -> Tuple[Tuple[bytes, ...], Tuple[bytes, ...]]:
+        raise NotImplementedError("BeaconChainDB classes must implement this method")
+
+    #
+    # Raw Database API
+    #
+    @abstractmethod
+    def exists(self, key: bytes) -> bool:
+        raise NotImplementedError("BeaconChainDB classes must implement this method")
+
+    @abstractmethod
+    def get(self, key: bytes) -> bytes:
+        raise NotImplementedError("BeaconChainDB classes must implement this method")
+
+    @abstractmethod
+    def persist_trie_data_dict(self, trie_data_dict: Dict[bytes, bytes]) -> None:
+        raise NotImplementedError("BeaconChainDB classes must implement this method")
+
+
+class BeaconChainDB(BaseBeaconChainDB):
+    db = None  # type: BaseAtomicDB
+
+    def __init__(self, db):
+        self.db = db
+
+    def persist_block(self,
+                      block: BaseBeaconBlock) -> Tuple[Tuple[bytes, ...], Tuple[bytes, ...]]:
+        """
+        Persist the given block.
+        """
+        with self.db.atomic_batch() as db:
+            return self._persist_block(db, block)
+
+    @classmethod
```

Comment:
What's the motivation for this being a `classmethod`?

Reply:
Copied our conversation:

> Piper: I think we did it so that we could implement some of them as staticmethod or classmethod
which the type system won't allow you to override those things cleanly

Reply:
The primary motivation for me is: we want to prevent accidental reads/writes of `self.db` in `_persist_block()`, which would break atomicity of `persist_block()`